lib/host/wasi/inode-win.cpp

```diff
@@ -82,53 +533,308 @@ INode INode::stdErr() noexcept {
   return INode(winapi::GetStdHandle(winapi::STD_ERROR_HANDLE_));
 }
 
-WasiExpect<INode> INode::open(std::string, __wasi_oflags_t, __wasi_fdflags_t,
-                              uint8_t) noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+WasiExpect<INode> INode::open(std::string Path, __wasi_oflags_t OpenFlags,
+                              __wasi_fdflags_t FdFlags,
+                              uint8_t VFSFlags) noexcept {
+
+  DWORD AttributeFlags = attributeFlags(OpenFlags, FdFlags);
+  DWORD AccessFlags = accessFlags(FdFlags, VFSFlags);
+  DWORD ShareFlags = shareFlags(OpenFlags);
+  DWORD CreationDisposition = creationDisposition(OpenFlags);
+
+  HANDLE FileHandle =
+      CreateFileA(Path.c_str(), AccessFlags, ShareFlags, nullptr,
+                  CreationDisposition, AttributeFlags, nullptr);
+
+  if (unlikely(FileHandle == INVALID_HANDLE_VALUE)) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  } else {
+    INode New(FileHandle);
+    return New;
+  }
 }
 
 WasiExpect<void> INode::fdAdvise(__wasi_filesize_t, __wasi_filesize_t,
                                  __wasi_advice_t) const noexcept {
+  // FIXME: No equivalent function was found for this purpose in the Win32 API
+  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+}
+
+WasiExpect<void> INode::fdAllocate(__wasi_filesize_t Offset,
+                                   __wasi_filesize_t Len) const noexcept {
+
+  LARGE_INTEGER FileSize;
+  if (unlikely(GetFileSizeEx(Handle, &FileSize) == FALSE)) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  // We need to check if the request size (Offset + Len) is lesser than the
+  // current the size and if it is lesser then we don't truncate the file
+
+  if (static_cast<int64_t>((Offset + Len) & 0x7FFFFFFFFFFFFFFF) >
+      FileSize.QuadPart) {
+
+    FILE_STANDARD_INFO StandardInfo;
+    FILE_ALLOCATION_INFO AllocationInfo;
+
+    if (unlikely(GetFileInformationByHandleEx(Handle, FileStandardInfo,
+                                              &StandardInfo,
+                                              sizeof(StandardInfo))) == FALSE)
+      return WasiUnexpect(fromWinError(GetLastError()));
+
+    // Update the size attribute
+    // TODO: Is it ok to do this?
```

Comment:
Sorry, what's the question?

Reply:
```
// Update the size attribute
// TODO: Is it ok to do this?
AllocationInfo.AllocationSize.QuadPart =
    static_cast<int64_t>((Offset + Len) & 0x7FFFFFFFFFFFFFFF);
```
We are converting a signed value into a unsigned value and here I am removing the sign bit. So the range of `QuadPart` is now the range of `uint64_t` (`Offset + Len`) / 2 which is lesser. So is this reduced range acceptable?

Reply:
I think you have to explain this in the comment.

Reply:
Done.